projects/ui/autocomplete/autocomplete-trigger.ts

```diff
@@ -0,0 +1,720 @@
+
+        return event
+          ? _getEventTarget<HTMLInputElement>(event)?.value
+          : this.host.nativeElement.value
+      }),
+    ),
+    { initialValue: null },
+  )
+
+  private readonly _windowBlur = toSignal(
+    fromEvent<FocusEvent>(this._document.defaultView ?? window, 'blur'),
+  )
+
+  private readonly _outsideClick = toSignal(
+    toObservable(this._overlayRef).pipe(
+      filter(Boolean),
+      switchMap(overlayRef => overlayRef.outsidePointerEvents()),
+      filter(event => {
+        const target = _getEventTarget<HTMLElement>(event)
+
+        return !this._origin().host.nativeElement.contains(target)
+      }),
+    ),
+  )
+
+  private readonly _overlayAttached = toSignal(
+    toObservable(this._overlayRef).pipe(
+      filter(Boolean),
+      switchMap(overlayRef => {
+        return merge(
+          overlayRef.attachments().pipe(startWith(true)),
+          overlayRef.detachments(),
+        ).pipe(map(() => overlayRef))
+      }),
+      map(overlayRef => overlayRef.hasAttached()),
+    ),
+    { initialValue: false },
+  )
+
+  private readonly _viewInitialized = new Observable(subscriber => {
+    afterNextRender(subscriber.next.bind(subscriber), {
+      injector: this._injector,
+    })
+  })
+
+  private readonly _optionsFirstRender = toSignal(
+    this._viewInitialized.pipe(
+      switchMap(() => {
+        return toObservable(this.autocomplete().options, {
+          injector: this._injector,
+        })
+      }),
+      filter(options => !!options.length),
+      take(1),
+    ),
+  )
+
+  constructor(
+    /** HTML Element of the trigger. */
+    public readonly host: ElementRef<HTMLInputElement>,
+    @Optional()
+    @Host()
+    private readonly _formField: CkFormField,
+    @Optional()
+    private readonly _directionality: Directionality,
+    @Optional()
+    @Inject(DOCUMENT)
+    private readonly _document: Document,
+    @Inject(CK_AUTOCOMPLETE_SCROLL_STRATEGY)
+    private readonly _scrollStrategy: ScrollStrategy,
+    private readonly _viewContainerRef: ViewContainerRef,
+    private readonly _overlay: Overlay,
+    private readonly _changeDetectorRef: ChangeDetectorRef,
+    private readonly _zone: NgZone,
+    private readonly _injector: Injector,
+  ) {
+    effect(this._windowBlurEffect.bind(this), { allowSignalWrites: true })
+    effect(this._originChangeEffect.bind(this))
+    effect(this._originWidthChangeEffect.bind(this))
+    effect(this._positionChangeEffect.bind(this))
+    effect(this._outsideClickEffect.bind(this))
+    effect(this._optionsFirstRenderEffect.bind(this))
+    effect(this._optionsChangesEffect.bind(this))
+    effect(this._animationOutDoneEffect.bind(this), { allowSignalWrites: true })
+    effect(this._optionsSelectionChangeEffect.bind(this), {
+      allowSignalWrites: true,
+    })
+  }
+
+  public ngOnDestroy(): void {
+    if (!this._overlayRef()) return
+
+    this.closePanel()
+    this._overlayRef()!.dispose()
```

Comment:
TS doesn't recognize the condition above. Unbelievable.

Reply:
Hehe. I was thinking about this one, and you know.. it makes perfect sense.. TS doesn't have signal data type. For TS that `this._overlayRef()` is a method, which might return value or `undefined` at any of its' calls. So no guarantee that you condition on line 294 will still be valid when interpreter reaches line 297 🤷‍♂️. So yea, maybe it's clearer to hold the value in a variable and operate with that variable instead of calling the signal all over the places